adapter/src/components/LoginModal.js

```diff
@@ -6,57 +6,133 @@ import {
+    password,
+    setError,
+    setIsLoggingIn,
+}) => {
+    try {
+        await postJSON(
+            `${server}/api/auth/login`,
+            JSON.stringify({
+                username: encodeURIComponent(username),
+                password: encodeURIComponent(password),
+            })
+        )
+        window.location.reload()
+    } catch (e) {
+        setError(e)
+        setIsLoggingIn(false)
+    }
+}
+
+const loginWithOldEndpoints = async ({ server, username, password }) => {
+    try {
+        await post(
+            `${server}/dhis-web-commons-security/login.action`,
+            `j_username=${encodeURIComponent(
+                username
+            )}&j_password=${encodeURIComponent(password)}`
+        )
+    } catch (e) {
+        console.error(e)
+    } finally {
+        window.location.reload()
+    }
+}
+
+export const LoginModal = ({ appName, baseUrl, loginApp = false }) => {
     const [server, setServer] = useState(baseUrl || '')
     const [username, setUsername] = useState('')
     const [password, setPassword] = useState('')
     const [isDirty, setIsDirty] = useState(false)
+    const [error, setError] = useState(null)
+    const [isLoggingIn, setIsLoggingIn] = useState(false)
 
     const isValid = (val) => val && val.length >= 2
+    const getSignInButtonText = ({ loginApp, isLoggingIn }) => {
+        if (!loginApp) {
+            return isLoggingIn ? i18n.t('Signing in...') : i18n.t('Sign in')
+        }
+        return isLoggingIn ? i18n.t('Going to app...') : i18n.t('Go to app')
+    }
 
     const onSubmit = async (e) => {
         e.preventDefault()
         setIsDirty(true)
-        if (isValid(server) && isValid(username) && isValid(password)) {
+        if (
+            isValid(server) &&
+            ((isValid(username) && isValid(password)) || loginApp)
+        ) {
+            setIsLoggingIn(true)
             if (!staticUrl) {
                 // keep the localStorage value here -- it's still used in some
                 // obscure cases, like in the cypress network shim
                 window.localStorage.DHIS2_BASE_URL = server
                 await setBaseUrlByAppName({ appName, baseUrl: server })
+                if (loginApp) {
+                    window.location.reload()
+                }
             }
-            try {
-                await post(
-                    `${server}/dhis-web-commons-security/login.action`,
-                    `j_username=${encodeURIComponent(
-                        username
-                    )}&j_password=${encodeURIComponent(password)}`
-                )
-            } catch (e) {
-                console.log(
-                    'TODO: This will always error and cancel the request until we get a real login endpoint!'
-                )
-            }
 
-            // TODO: Hacky solution... this shouldn't require a reload
-            window.location.reload()
+            const useNewLoginAPI = await getUseNewLoginAPI(server)
```

Comment:
This feels like it's clashing a bit with the react hook naming conventions -- what do you think about `isNewLoginAPIAvailable` and `getIsNewLoginAPIAvailable`?